### DISCUSSION TITLE
Fichier layout non trouvé sans affichage d'erreur

### DISCUSSION THREAD

**Discussion by moimp3** [rawsrc/PhpEcho#29] [Q&A] [answered]
Bonjour,
J'ai un fichier /user/index.php qui se termine comme ceci:
```php
var_dump(DIR_ROOT.'user'.DIRECTORY_SEPARATOR.'view');	// renvoie 'C:\wamp64\www\proginet\appOmnes\user\view'
PhpEcho::setTemplateDirRoot(DIR_ROOT.'user'.DIRECTORY_SEPARATOR.'view');

// start application
####################

$page = $_GET['forgot'] ?? '';

require_once USER_VIEW_PAGE.'header.php';
require_once USER_VIEW_PAGE.'footer.php';

switch($page){
	case 'password':
		require_once USER_VIEW_PAGE.'forgotPassword.php';
		break;
	case 'login':
		require_once USER_VIEW_PAGE.'forgotLogin.php';
		break;
}
```
Le fichier /user/page/forgotPassword.php contient le code suivant:
```php
<?php declare(strict_types=1);
// FORGOT PASSWORD

use rawsrc\PhpEcho\PhpEcho;

$pw					= new PhpEcho('layout/main.php');

$pw['title']		= $textes->head_title;
$pw['description']	= $textes->head_description;
$pw['language']		= $language;

$body				= new PhpEcho('block/forgotPassword.php');
$body['errors']		= writeAlerts();
$body['infos']		= writeAlerts(1);
$body['language']	= $language;

$body['h2']			= $textes->pwd_h2;
$body['login']		= $textes->pwd_login;
$body['email']		= $textes->pwd_email;
$body['firstname']	= $textes->user_firstname;
$body['familyname']	= $textes->user_familyname;
$body['goOn']		= $textes->btn_goOn;
$body['back']		= $textes->btn_back;


// for next two lines, pages are required in index.php and parameters are defined in relevant pages 
$body['header']		= $header;
$body['footer']		= $footer;

$pw['body']			= $body;

echo $body;
```
Bizarrement, les blocs sont trouvés mais le fichier /user/view/layout/main.php n'est pas trouvé. Pourtant aucune erreur ne s'affiche.

**Comment by rawsrc** [comment]
Salut,

Il n'y aurait pas un souci dans le code du fichier main.php à tout hasard ?
Généralement une page blanche = crash de php

rawsrc

**Comment by moimp3** [comment]
J'ai une application principale (1) à la racine du site et une gestion des utilisateurs (2) dans /user.
Les deux sont indépendante et ont chacune un fichier index dans lequel est défini `setTemplateDirRoot`. Les deux sont indépendantes mais (2) est cependant appelées à partir de (1). Je me demande s'il n'y aurait pas un conflit.
Lorsque j'affiche le code source (ctrl+U), je n'ai aucune trace de l'entête du document défini dans `/user/view/layout/main.php` par contre le corps de page (balise body) s'affiche intégralement.
Si je supprime le fichier `/user/view/layout/main.php`, je ne provoque aucune erreur.

P.S. Je note une contradiction dans README dans le "Répertoire racine de toutes les vues". Tu écris 

> Veuillez noter que le seul séparateur de répertoire autorisé est / (slash).

et plus bas dans l'exemple tu emploies DIRECTORY_SEPARATOR.

**Comment by rawsrc** [comment]
Salut,

Tu dis que c'est indépendant mais `root/index.php` appelle `root/user/index.php `
Les deux fichiers `index.php` font appel à `PhpEcho::setTemplateDirRoot()`, c'est le dernier qui a parlé qui a raison (c'est une fonction statique avec une portée globale commune à toutes les instances). Donc d'après ton déroulé c'est le `PhpEcho::setTemplateDirRoot()` inclus dans `root/user/index.php ` qui prime. 

Dans ton code, à l'endroit où tu penses que ça déconne, regarde le retour de `PhpEcho::getFullFilepath('layout/main.php);` et vérifie le chemin complet du fichier tel que généré par le moteur.

Il n'y a pas de contradiction : dans le passage de paramètre pour ce qui est des noms de fichiers, le seul séparateur autorisé est `/`. Ensuite le moteur adapte le chemin résolu en fonction du système et donc il se peut que `/` devienne '\\' par l'utilisation de `DIRECTORY_SEPARATOR`.

**Comment by moimp3** [comment]
Tout d'abord merci pour le `PhpEcho::getFullFilepath('layout/main.php);`.
La valeur retournée par ce getFullFilepath est correcte ('C:\wamp64\www\proginet\appOmnes\user\view\layout\main.php') mais le fichier n'est pas lu (A titre d'essai le fichier ne contient en tout et pour tout que la ligne `echo 'xxxx';` qui ne s'affiche pas.)

**Comment by rawsrc** [comment, marked as answer]
tu fais bien 
```php
PhpEcho::setTemplateDirRoot("ta_valeur");
echo new PhpEcho('layout/main.php');
exit;
```
Et rien ne s'affiche ?

**Comment by moimp3** [comment]
Ouf !!, Merci. J'ai du passer au moins 10 fois sur mon erreur sans la voir. J'avais:
``
$pw = new PhpEcho('layout/main.php');
echo $uneAutreVariable;
``
C'est sûr que ça ne pouvait pas marcher.
J'avais déjà remarqué ce phénomène: Plus on se relis, moins on voit ses erreurs. Si on lit le travail d'un autre (ou la sienne plusieurs jours après), l'erreur nous saute aux yeux.

**Comment by rawsrc** [comment]
ben ben voilà, c'est résolu
la prochaine fois demande avant  à ta femme 😉😊

@++
